assert len(indices) == self.total_size error during multiple GPU training

I am trying to train my dataset on 8 GPU's. However, after calling `./dist_train.sh` this error assertion appeares:

Traceback (most recent call last):                
  File "./tools/train.py", line 113, in <module>  
    main()               
  File "./tools/train.py", line 109, in main      
    logger=logger)       
  File "/mmdetection/mmdet/apis/train.py", line 58, in train_detector                                
    _dist_train(model, dataset, cfg, validate=validate)                                              
  File "/mmdetection/mmdet/apis/train.py", line 186, in _dist_train                                  
    runner.run(data_loaders, cfg.workflow, cfg.total_epochs)                                         
  File "/opt/conda/lib/python3.6/site-packages/mmcv/runner/runner.py", line 358, in run
    epoch_runner(data_loaders[i], **kwargs)       
  File "/opt/conda/lib/python3.6/site-packages/mmcv/runner/runner.py", line 260, in train
    for i, data_batch in enumerate(data_loader):  
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 193, in __iter__    return _DataLoaderIter(self)                  
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 493, in __init__
    self._put_indices()  
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 591, in _put_indices
    indices = next(self.sample_iter, None)        
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/sampler.py", line 172, in __iter__
    for idx in self.sampler:                      
  File "/mmdetection/mmdet/datasets/loader/sampler.py", line 138, in __iter__                        
    assert len(indices) == self.total_size        
...

in the config I tried various values for `imgs_per_gpu` and `workers_per_gpu`, currently it is:
`imgs_per_gpu=2,
workers_per_gpu=2,`
no settings was working though. Single-GPU training works well.

What is the meaning of this assert?
Thanks!

Hello!

I have situation, when 6 pics, where w>h.
33994 pics, when h>w.
And 2 pics, where h==w.

I have deleted h==w pics and have AssertionError: assert len(indices) == self.total_size anyway :(

--- a/mmdet/datasets/loader/sampler.py
+++ b/mmdet/datasets/loader/sampler.py
@@ -132,8 +132,12 @@ class DistributedGroupSampler(Sampler):
                     math.ceil(
                         size * 1.0 / self.samples_per_gpu / self.num_replicas)
                 ) * self.samples_per_gpu * self.num_replicas - len(indice)
-                indice += indice[:extra]
-                indices += indice
+                # pad indice
+                tmp = indice.copy()
+                for _ in range(extra//size):
+                    indice.extend(tmp)
+                indice.extend(tmp[:extra%size])
+                indices.extend(indice)
 
         assert len(indices) == self.total_size
 

--- a/mmdet/datasets/loader/sampler.py
+++ b/mmdet/datasets/loader/sampler.py
@@ -134,9 +134,9 @@ class DistributedGroupSampler(Sampler):
                 ) * self.samples_per_gpu * self.num_replicas - len(indice)
                 # pad indice
                 tmp = indice.copy()
-                for _ in range(extra//size):
+                for _ in range(extra // size):
                     indice.extend(tmp)
-                indice.extend(tmp[:extra%size])
+                indice.extend(tmp[:extra % size])
                 indices.extend(indice)
 
         assert len(indices) == self.total_size

